docs/software/ml/pytorch.md

```diff
@@ -0,0 +1,394 @@
+        3. The MPI compiler wrappers are also available.
+        4. Exit the uenv.
+
+=== "Spack"
+
+    The pytorch uenv can also be used as a base for building software with Spack, because it provides compilers, MPI, Python and common packages like HDF5.
+
+    [Check out the guide for using Spack with uenv][ref-building-uenv-spack].
+
+[](){#ref-uenv-pytorch-venv}
+## Adding Python packages on top of the uenv
+
+Uenvs are read-only, and cannot be modified. However, it is possible to add Python packages on top of the uenv using virtual environments.
+
+!!! example "creating a virtual environment on top of the uenv"
+    ```console
+    $ uenv start pytorch/v2.6.0:v1 --view=default # (1)!
+
+    $ python -m venv ./my-venv # (2)!
+
+    $ source ./my-venv/bin/activate # (3)!
+
+    (my-venv) $ pip install <package> # (4)!
+
+    (my-venv) $ deactivate # (5)!
+
+    $ exit # (6)!
+    ```
+
+    1. The `default` view is recommended, as it loads all the packages provided by the uenv.
+       This is important for PyTorch to work correctly, as it relies on the CUDA and NCCL libraries provided by the uenv.
+    2. The virtual environment is created in the current working directory, and can be activated and deactivated like any other Python virtual environment.
+    3. Activating the virtual environment will override the Python executable provided by the uenv, and use the one from the virtual environment instead.
+       This is important to ensure that the packages installed in the virtual environment are used.
+    4. The virtual environment can be used to install any Python package.
+    5. The virtual environment can be deactivated using the `deactivate` command.
+       This will restore the original Python executable provided by the uenv.
+    6. The uenv can be exited using the `exit` command or by typing `ctrl-d`.
+
+!!! note
+    Python virtual environments can be slow on the parallel Lustre file system due to the amount of small files and potentially many processes accessing it.
+    If this becomes a bottleneck, consider [squashing the venv][ref-guides-storage-venv] into its own memory-mapped, read-only file system to enhance scalability and reduce load times.
+
+Alternatively one can use the uenv as [upstream Spack instance][ref-building-uenv-spack] to to add both Python and non-Python packages.
+However, this workflow is more involved and intended for advanced Spack users.
+
+## Running PyTorch jobs with SLURM
+
+!!! example "slurm sbatch script"
+    ```bash
+    #!/bin/bash
+    #SBATCH --job-name=myjob
+    #SBATCH --nodes=1
+    #SBATCH --ntasks-per-node=4
+    #SBATCH --cpus-per-task=72
+    #SBATCH --time=00:30:00
+    # (1)!
+    #SBATCH --uenv=pytorch/v2.6.0:/user-environment
+    #SBATCH --view=default
+
+    #################################
+    # OpenMP environment variables #
+    #################################
+    export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK  # (2)!
```

Comment:
```suggestion
    export OMP_NUM_THREADS=8  # (2)!
```
so as we are talking about pytorch, if they are using the dataloader or anything else that forks the processes, this number should be +- divided by the number of forks. But they should really test what work best for them? I bet 72 is probably too much...

Reply:
- We can't know if anything uses openmp in  the first place
- maybe 72 is too much indeed
I'll reduce it and comment on the choice in the annotation

Reply:
torch uses that for cpu ops as well (and so does numpy AFAIK)
```
OMP_NUM_THREADS=42 python -c 'import torch; print(torch.get_num_threads())'
42
```